File: 354-russian-doll-envelopes/russian-doll-envelopes.py
class Solution:
    def maxEnvelopes(self, envelopes: List[List[int]]) -> int:
        """
        order by widths: [2, 3] [5, 4] [6, 4] [6, 7]
        order by heights: [2, 3] [5, 4] [6, 4] [6, 7]

        - essentially need to find LIS where widths & height both strictly increasing 
        - but how do we combine two dimensions?

        within same width group only 1 envelope can be chosen in the LIS of heights
            - order by width ASC, height DESC to ensure this
        
        [1, 3, 2, 5]

        seq = [1, 2, 3]
                 
        """
        envelopes.sort(key = lambda x : (x[0], -x[1]))

        seq = []

        for _, height in envelopes:
            if not seq or seq[-1] < height:
                seq.append(height)
                continue
            
            i, j = 0, len(seq) - 1
            while i <= j:
                mid = (i + j) // 2
                if seq[mid] < height:
                    i = mid + 1
                else:
                    j = mid - 1

            seq[i] = height
        
        return len(seq)

File: 354-russian-doll-envelopes/russian-doll-envelopes.go
/*
(2, 3) (5, 4) (6, 5) (6, 7)

3 5 6
         4
*/

import (
    "slices"
    "cmp"
    "fmt"
)

func maxEnvelopes(envelopes [][]int) int {
    slices.SortFunc(envelopes, func(x, y []int) int {
        if x[0] != y[0] {
            return cmp.Compare(x[0], y[0])
        }
        // descending height when widths equal
        return cmp.Compare(y[1], x[1])
    })

    seq := make([]int, 0)
    for i := range(len(envelopes)) {
        if len(seq) == 0 || envelopes[i][1] > seq[len(seq) - 1]{
            // empty or increasing seq
            seq = append(seq, envelopes[i][1])
        }

        low := 0
        high := len(seq) - 1

        for low <= high{
            mid := low + (high - low) / 2
            if seq[mid] >= envelopes[i][1]{
                high = mid - 1
            } else {
                low = mid + 1
            }
        }
        seq[low] = envelopes[i][1]
    }

    return len(seq)
}